File: 2300.java
import java.util.Arrays;

class Solution {

    private int lowerBound(int[] arr, int key) {
        int low = 0, high = arr.length;
        while (low < high) {
            int middle = low + (high - low) / 2;
            if (arr[middle] < key) {
                low = middle + 1;
            } else {
                high = middle;
            }
        }
        return low;
    }

    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int[] result = new int[spells.length];
        Arrays.sort(potions);
        int potlen = potions.length;
        int maxPotion = potions[potlen - 1];
        for (int i = 0; i < spells.length; i++) {
            int spell = spells[i];
            long minPotion = (long) Math.ceil((1.0 * success) / spell);
            if (minPotion > maxPotion) {
                result[i] = 0;
                continue;
            }
            int index = lowerBound(potions, (int) minPotion);
            result[i] = potlen - index;
        }
        return result;
    }
}

File: Python/2300.py
from typing import List


class Solution:
    def successfulPairs(
        self, spells: List[int], potions: List[int], success: int
    ) -> List[int]:
        potions.sort()
        n, m = len(spells), len(potions)
        result = [0] * n
        for i, spell in enumerate(spells):
            left, right = 0, m
            while left < right:
                mid = left + (right - left) // 2
                if spell * potions[mid] >= success:
                    right = mid
                else:
                    left = mid + 1
            result[i] = m - left
        return result
